### ISSUE TITLE
Use generators with different topologies?

### ISSUE THREAD

**Issue by jeiros** [Acellera/htmd#367]
Is it possible to have an adaptive sampling scheme which uses different topologies for each generator?  

I have a protein which I've docked a ligand to in different positions. I've built the system directly with tleap. The `solvatebox` command in tleap adds a different amount of water molecules every time since the position of the ligand affects where the farthest point from the centre of the system is (also one of my generators has the ligand unbound, placed 1 nm away from the protein).

So I have topologies with a different amount of water molecules. The system is exactly the same, once the waters and the counterions are stripped.

I've been able to start the adaptive simulations with the following commands:

```python
import htmd
from htmd import Molecule
import shutil
import os
from glob import glob
from htmd.protocols.pmemdproduction import Production
from htmd.projections.metricdistance import MetricDistance
app = htmd.LocalGPUQueue()
app.ngpu = 4
app.datadir = './data'
app.trajext = 'nc'

adapt = htmd.AdaptiveMD()
adapt.nmin = 3
adapt.nmax = 4
adapt.nepochs = 20
adapt.updateperiod = 2 * 3600  # Every 2 hours
adapt.projection = MetricDistance(sel1='resname SIB and name C1',  #  C1 carbon
                                  sel2='name CA')
adapt.app = app
adapt.filtersel = 'not water and not resname "Na\+" "Cl\-"'
adapt.coorname = 'structure.ncrst'
adapt.run()
```
with a pre-build directory structure like this one:
```
.
├── [  50]  generators
│   ├── [  99]  gen1
│   │   ├── [ 241]  Production.in
│   │   ├── [ 163]  run.sh
│   │   ├── [ 11M]  structure.ncrst
│   │   └── [ 39M]  structure.prmtop
│   ├── [  99]  gen2
│   │   ├── [ 241]  Production.in
│   │   ├── [ 163]  run.sh
│   │   ├── [ 11M]  structure.ncrst
│   │   └── [ 39M]  structure.prmtop
│   ├── [  99]  gen3
│   │   ├── [ 241]  Production.in
│   │   ├── [ 163]  run.sh
│   │   ├── [ 11M]  structure.ncrst
│   │   └── [ 39M]  structure.prmtop
│   └── [  99]  gen4
│       ├── [ 241]  Production.in
│       ├── [ 163]  run.sh
│       ├── [ 13M]  structure.ncrst
│       └── [ 45M]  structure.prmtop
├── [2.3K]  run-adaptive.py
├── [2.8M]  stripped.prmtop
├── [   0]  structure.ncrst
└── [   0]  structure.prmtop
```
where each `structure.prmtop` and `structure.ncrst` files in the generator subfolders are different. I 'hacked' the adaptive algorithm to start and be able to build a simlist by creating the empty `structure.ncrst` and `structure.prmtop` files in the base folder.

Now the first epoch simulations are starting to finish and I think the adaptive code is having issues doing the filtering and knowing which `structure.ncrst` file to use.

First I got a long traceback ending with a multiprocessing error:

<details>

```
2017-06-22 03:15:20,368 - htmd.adaptive.adaptive - INFO - Sleeping for 7200 seconds.
2017-06-22 04:28:05,525 - htmd.queues.localqueue - INFO - Completed /home/je714/G159D_P/SilybinB/adaptive/S1P/input/e1s1_gen1
2017-06-22 05:15:20,468 - htmd.adaptive.adaptive - INFO - Processing epoch 1
2017-06-22 05:15:20,468 - htmd.adaptive.adaptive - INFO - Retrieving simulations.
2017-06-22 05:15:20,469 - htmd.adaptive.adaptive - INFO - 3 simulations in progress
2017-06-22 05:15:20,469 - htmd.adaptive.adaptiverun - INFO - Postprocessing new data

HTMD License accepted automatically. Check license here: https://raw.githubusercontent.com/Acellera/htmd/master/htmd/LICENCE.txt

For advanced features (e.g. parameterize) and to remove this message, we recommend registering. Run htmd_register in your terminal.

Please cite HTMD: Doerr et al.(2016)JCTC,12,1845. https://dx.doi.org/10.1021/acs.jctc.6b00049

HTMD Documentation at: https://www.htmd.org/docs/latest/

You are on the latest HTMD version (unpackaged : /home/je714/htmd/htmd).


2017-06-22 05:21:58,250 - htmd.molecule.writers - WARNING - Field "serial" of PDB overflows. Your data will be truncated to 5 characters.
2017-06-22 05:21:59,335 - htmd.molecule.writers - WARNING - Field "resid" of PDB overflows. Your data will be truncated to 4 characters.

2017-06-22 05:26:50,403 - htmd.projections.metric - INFO - Frame step 200.0ns was read from the trajectories. If it looks wrong, redefine it by manually setting the MetricData.fstep property.
2017-06-22 05:26:50,404 - htmd.metricdata - INFO - Dropped 0 trajectories from 1 resulting in 1
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/pyemma/__init__.py:91: UserWarning: You are not using the latest release of PyEMMA. Latest is 2.4, you have 2.3.2.
  .format(latest=latest, current=current), category=UserWarning)
2017-06-22 05:26:51,568 - htmd.metricdata - INFO - Dropped 0 trajectories from 1 resulting in 1
2017-06-22 05:26:54,591 - htmd.model - INFO - 100.0% of the data was used
2017-06-22 05:26:54,591 - htmd.model - INFO - Number of trajectories that visited each macrostate:
2017-06-22 05:26:54,592 - htmd.model - INFO - [1 1 1 1 1 1 1 1]
2017-06-22 05:26:54,592 - htmd.model - INFO - Take care! Macro 0 has been visited only in 1 trajectories:
2017-06-22 05:26:54,592 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 1 has been visited only in 1 trajectories:
2017-06-22 05:26:54,593 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 2 has been visited only in 1 trajectories:
2017-06-22 05:26:54,593 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 3 has been visited only in 1 trajectories:
2017-06-22 05:26:54,593 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 4 has been visited only in 1 trajectories:
2017-06-22 05:26:54,593 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 5 has been visited only in 1 trajectories:
2017-06-22 05:26:54,593 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,593 - htmd.model - INFO - Take care! Macro 6 has been visited only in 1 trajectories:
2017-06-22 05:26:54,594 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]

2017-06-22 05:26:54,594 - htmd.model - INFO - Take care! Macro 7 has been visited only in 1 trajectories:
2017-06-22 05:26:54,594 - htmd.model - INFO - 
simid = 0
parent = 0
input = None
trajectory = ['filtered/e1s1_gen1/Production.filtered.nc']
molfile = filtered/filtered.pdb
numframes = [None]



2017-06-22 05:27:05,295 - htmd.molecule.writers - WARNING - Field "serial" of PDB overflows. Your data will be truncated to 5 characters.
2017-06-22 05:27:06,413 - htmd.molecule.writers - WARNING - Field "resid" of PDB overflows. Your data will be truncated to 4 characters.
2017-06-22 05:31:33,515 - htmd.molecule.writers - WARNING - Field "serial" of PDB overflows. Your data will be truncated to 5 characters.
2017-06-22 05:31:34,594 - htmd.molecule.writers - WARNING - Field "resid" of PDB overflows. Your data will be truncated to 4 characters.

multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/je714/htmd/htmd/molecule/writers.py", line 461, in MDTRAJwrite
    mol.write(tmpxtc)
  File "/home/je714/htmd/htmd/molecule/molecule.py", line 1149, in write
    _WRITERS[ext](src, filename)
  File "/home/je714/htmd/htmd/molecule/writers.py", line 235, in XTCwrite
    cstep = ct.c_int(step[f])
IndexError: invalid index to scalar variable.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 350, in __call__
    return self.func(*args, **kwargs)
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py", line 131, in __call__
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py", line 131, in <listcomp>
    return [func(*args, **kwargs) for func, args, kwargs in self.items]
  File "/home/je714/htmd/htmd/adaptive/adaptive.py", line 245, in _writeInputsFunction
    mol.write(path.join(newDir, coorname))
  File "/home/je714/htmd/htmd/molecule/molecule.py", line 1149, in write
    _WRITERS[ext](src, filename)
  File "/home/je714/htmd/htmd/molecule/writers.py", line 472, in MDTRAJwrite
    raise ValueError('MDtraj reader failed for file {} with error "{}"'.format(filename, e))
ValueError: MDtraj reader failed for file input/e2s1_e1s1p0f607/structure.ncrst with error "invalid index to scalar variable."

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 359, in __call__
    raise TransportableException(text, e_type)
joblib.my_exceptions.TransportableException: TransportableException
___________________________________________________________________________
ValueError                                         Thu Jun 22 05:35:20 2017
PID: 7947    Python 3.6.1: /home/je714/miniconda3/envs/htmd-py36/bin/python
...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in __call__(self=<joblib.parallel.BatchedCalls object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        self.items = [(<function _writeInputsFunction>, (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst'), {})]
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in <listcomp>(.0=<list_iterator object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        func = <function _writeInputsFunction>
        args = (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst')
        kwargs = {}
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptive.py in _writeInputsFunction(i=0, f=sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, epoch=2, inputpath='input', coorname='structure.ncrst')
    240         mol = Molecule(currSim.molfile)
    241     else:
    242         mol = Molecule()
    243     mol.read(traj)
    244     mol.dropFrames(keep=frameNum)  # Making sure only specific frame to write is kept
--> 245     mol.write(path.join(newDir, coorname))
        mol.write = <bound method Molecule.write of <htmd.molecule.m...gen1/structure.prmtop
viewname: structure.prmtop>
        newDir = 'input/e2s1_e1s1p0f607/'
        coorname = 'structure.ncrst'
    246 
    247 
    248 def epochSimIndexes(simlist):
    249     """ Finds the simulation indexes for each epoch.

...........................................................................
/home/je714/htmd/htmd/molecule/molecule.py in write(self=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst', sel=None, type=None)
   1144             src.filter(sel, _logger=False)
   1145 
   1146         if type in _WRITERS:
   1147             ext = type
   1148         if ext in _WRITERS:
-> 1149             _WRITERS[ext](src, filename)
        _WRITERS = {'binpos': <function MDTRAJwrite>, 'coor': <function BINCOORwrite>, 'crd': <function MDTRAJwrite>, 'dcd': <function MDTRAJwrite>, 'gro': <function GROwrite>, 'h5': <function MDTRAJwrite>, 'lammpstrj': <function MDTRAJwrite>, 'lh5': <function MDTRAJwrite>, 'mdcrd': <function MDTRAJwrite>, 'mol2': <function MOL2write>, ...}
        ext = 'ncrst'
        src = <htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
   1150 
   1151     def empty(self, numAtoms):
   1152         """ Creates an empty molecule of N atoms.
   1153 

...........................................................................
/home/je714/htmd/htmd/molecule/writers.py in MDTRAJwrite(mol=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst')
    467         # traj.xyz = np.swapaxes(np.swapaxes(self.coords, 1, 2), 0, 1) / 10
    468         # traj.time = self.time
    469         # traj.unitcell_lengths = self.box.T / 10
    470         traj.save(filename)
    471     except Exception as e:
--> 472         raise ValueError('MDtraj reader failed for file {} with error "{}"'.format(filename, e))
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
        e = undefined
    473 
    474 
    475 _WRITERS = {'psf': PSFwrite,
    476             'pdb': PDBwrite,

ValueError: MDtraj reader failed for file input/e2s1_e1s1p0f607/structure.ncrst with error "invalid index to scalar variable."
___________________________________________________________________________
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py", line 699, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/multiprocessing/pool.py", line 608, in get
    raise self._value
joblib.my_exceptions.TransportableException: TransportableException
___________________________________________________________________________
ValueError                                         Thu Jun 22 05:35:20 2017
PID: 7947    Python 3.6.1: /home/je714/miniconda3/envs/htmd-py36/bin/python
...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in __call__(self=<joblib.parallel.BatchedCalls object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        self.items = [(<function _writeInputsFunction>, (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst'), {})]
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in <listcomp>(.0=<list_iterator object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        func = <function _writeInputsFunction>
        args = (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst')
        kwargs = {}
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptive.py in _writeInputsFunction(i=0, f=sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, epoch=2, inputpath='input', coorname='structure.ncrst')
    240         mol = Molecule(currSim.molfile)
    241     else:
    242         mol = Molecule()
    243     mol.read(traj)
    244     mol.dropFrames(keep=frameNum)  # Making sure only specific frame to write is kept
--> 245     mol.write(path.join(newDir, coorname))
        mol.write = <bound method Molecule.write of <htmd.molecule.m...gen1/structure.prmtop
viewname: structure.prmtop>
        newDir = 'input/e2s1_e1s1p0f607/'
        coorname = 'structure.ncrst'
    246 
    247 
    248 def epochSimIndexes(simlist):
    249     """ Finds the simulation indexes for each epoch.

...........................................................................
/home/je714/htmd/htmd/molecule/molecule.py in write(self=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst', sel=None, type=None)
   1144             src.filter(sel, _logger=False)
   1145 
   1146         if type in _WRITERS:
   1147             ext = type
   1148         if ext in _WRITERS:
-> 1149             _WRITERS[ext](src, filename)
        _WRITERS = {'binpos': <function MDTRAJwrite>, 'coor': <function BINCOORwrite>, 'crd': <function MDTRAJwrite>, 'dcd': <function MDTRAJwrite>, 'gro': <function GROwrite>, 'h5': <function MDTRAJwrite>, 'lammpstrj': <function MDTRAJwrite>, 'lh5': <function MDTRAJwrite>, 'mdcrd': <function MDTRAJwrite>, 'mol2': <function MOL2write>, ...}
        ext = 'ncrst'
        src = <htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
   1150 
   1151     def empty(self, numAtoms):
   1152         """ Creates an empty molecule of N atoms.
   1153 

...........................................................................
/home/je714/htmd/htmd/molecule/writers.py in MDTRAJwrite(mol=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst')
    467         # traj.xyz = np.swapaxes(np.swapaxes(self.coords, 1, 2), 0, 1) / 10
    468         # traj.time = self.time
    469         # traj.unitcell_lengths = self.box.T / 10
    470         traj.save(filename)
    471     except Exception as e:
--> 472         raise ValueError('MDtraj reader failed for file {} with error "{}"'.format(filename, e))
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
        e = undefined
    473 
    474 
    475 _WRITERS = {'psf': PSFwrite,
    476             'pdb': PDBwrite,

ValueError: MDtraj reader failed for file input/e2s1_e1s1p0f607/structure.ncrst with error "invalid index to scalar variable."
___________________________________________________________________________

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "run-adaptive.py", line 70, in <module>
    adapt.run()
  File "/home/je714/htmd/htmd/adaptive/adaptive.py", line 99, in run
    flag = self._algorithm()
  File "/home/je714/htmd/htmd/adaptive/adaptiverun.py", line 133, in _algorithm
    self._writeInputs(self._model.data.rel2sim(np.concatenate(relFrames)))
  File "/home/je714/htmd/htmd/adaptive/adaptive.py", line 203, in _writeInputs
    delayed(_writeInputsFunction)(i, f, epoch, self.inputpath, self.coorname) for i, f in enumerate(simsframes))
  File "/home/je714/htmd/htmd/parallelprogress.py", line 44, in tmp
    return Parallel(**joblib_args)(bar_func(op_iter))
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py", line 789, in __call__
    self.retrieve()
  File "/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py", line 740, in retrieve
    raise exception
joblib.my_exceptions.JoblibValueError: JoblibValueError
___________________________________________________________________________
Multiprocessing exception:
...........................................................................
/home/je714/G159D_P/SilybinB/adaptive/S1P/run-adaptive.py in <module>()
     65 adapt.projection = MetricDistance(sel1='resname SIB and name C1',  # hydroxils and C1 carbon
     66                                   sel2='name CA')
     67 adapt.app = app
     68 adapt.filtersel = 'not water and not resname "Na\+" "Cl\-"'
     69 adapt.coorname = 'structure.ncrst'
---> 70 adapt.run()

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptive.py in run(self=<htmd.adaptive.adaptiverun.AdaptiveMD object>)
     94                     self._unsetLock()
     95                     return
     96 
     97                 # If currently running simulations are lower than nmin start new ones to reach nmax number of sims
     98                 if self._running <= self.nmin and epoch < self.nepochs:
---> 99                     flag = self._algorithm()
        flag = undefined
        self._algorithm = <bound method AdaptiveMD._algorithm of <htmd.adaptive.adaptiverun.AdaptiveMD object>>
    100                     if flag is False:
    101                         self._unsetLock()
    102                         return
    103 

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptiverun.py in _algorithm(self=<htmd.adaptive.adaptiverun.AdaptiveMD object>)
    128 
    129         N = self.nmax - self._running
    130         reward = self._criteria(self._model, self.method)
    131         reward = self._truncate(reward, N)
    132         relFrames, _, _ = self._getSpawnFrames(reward, self._model, self._model.data, N)
--> 133         self._writeInputs(self._model.data.rel2sim(np.concatenate(relFrames)))
        self._writeInputs = <bound method AdaptiveBase._writeInputs of <htmd.adaptive.adaptiverun.AdaptiveMD object>>
        self._model.data.rel2sim = <bound method MetricData.rel2sim of <htmd.metric...ata'> at 0x7f75bcf642e8
trajectories shape: (1,)>
        relFrames = [array([[  0, 607]])]
    134         return True
    135 
    136     def _checkNFrames(self, data):
    137         if self.nframes != 0 and data.numFrames >= self.nframes:

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptive.py in _writeInputs(self=<htmd.adaptive.adaptiverun.AdaptiveMD object>, simsframes=array([ sim = 
simid = 0
parent = 0
input = None...s = [None]

piece = 0
frame = 607], dtype=object), epoch=2)
    198 
    199         from htmd.parallelprogress import ParallelExecutor
    200         from htmd.config import _config
    201         aprun = ParallelExecutor(n_jobs=_config['ncpus'])
    202         aprun(total=len(simsframes), description='Writing inputs')(
--> 203             delayed(_writeInputsFunction)(i, f, epoch, self.inputpath, self.coorname) for i, f in enumerate(simsframes))
        epoch = 2
        self.inputpath = 'input'
        self.coorname = 'structure.ncrst'
        simsframes = array([ sim = 
simid = 0
parent = 0
input = None...s = [None]

piece = 0
frame = 607], dtype=object)
    204 
    205     @abc.abstractmethod
    206     def _algorithm(self):
    207         return

...........................................................................
/home/je714/htmd/htmd/parallelprogress.py in tmp(op_iter=<generator object AdaptiveBase._writeInputs.<locals>.<genexpr>>)
     39     """
     40     def aprun(**tq_args):
     41         def tmp(op_iter):
     42             foo = lambda args: lambda x: progressbar(x, **args)
     43             bar_func = foo(tq_args)
---> 44             return Parallel(**joblib_args)(bar_func(op_iter))
        bar_func = <function ParallelExecutor.<locals>.aprun.<locals>.tmp.<locals>.<lambda>.<locals>.<lambda>>
        op_iter = <generator object AdaptiveBase._writeInputs.<locals>.<genexpr>>
     45         return tmp
     46     return aprun

...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in __call__(self=Parallel(n_jobs=-2), iterable=<generator object progressbar>)
    784             if pre_dispatch == "all" or n_jobs == 1:
    785                 # The iterable was consumed all at once by the above for loop.
    786                 # No need to wait for async callbacks to trigger to
    787                 # consumption.
    788                 self._iterating = False
--> 789             self.retrieve()
        self.retrieve = <bound method Parallel.retrieve of Parallel(n_jobs=-2)>
    790             # Make sure that we get a last message telling us we are done
    791             elapsed_time = time.time() - self._start_time
    792             self._print('Done %3i out of %3i | elapsed: %s finished',
    793                         (len(self._output), len(self._output),

---------------------------------------------------------------------------
Sub-process traceback:
---------------------------------------------------------------------------
ValueError                                         Thu Jun 22 05:35:20 2017
PID: 7947    Python 3.6.1: /home/je714/miniconda3/envs/htmd-py36/bin/python
...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in __call__(self=<joblib.parallel.BatchedCalls object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        self.items = [(<function _writeInputsFunction>, (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst'), {})]
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/miniconda3/envs/htmd-py36/lib/python3.6/site-packages/joblib/parallel.py in <listcomp>(.0=<list_iterator object>)
    126     def __init__(self, iterator_slice):
    127         self.items = list(iterator_slice)
    128         self._size = len(self.items)
    129 
    130     def __call__(self):
--> 131         return [func(*args, **kwargs) for func, args, kwargs in self.items]
        func = <function _writeInputsFunction>
        args = (0, sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, 2, 'input', 'structure.ncrst')
        kwargs = {}
    132 
    133     def __len__(self):
    134         return self._size
    135 

...........................................................................
/home/je714/htmd/htmd/adaptive/adaptive.py in _writeInputsFunction(i=0, f=sim = 
simid = 0
parent = 0
input = None
traject...red.pdb
numframes = [None]

piece = 0
frame = 607, epoch=2, inputpath='input', coorname='structure.ncrst')
    240         mol = Molecule(currSim.molfile)
    241     else:
    242         mol = Molecule()
    243     mol.read(traj)
    244     mol.dropFrames(keep=frameNum)  # Making sure only specific frame to write is kept
--> 245     mol.write(path.join(newDir, coorname))
        mol.write = <bound method Molecule.write of <htmd.molecule.m...gen1/structure.prmtop
viewname: structure.prmtop>
        newDir = 'input/e2s1_e1s1p0f607/'
        coorname = 'structure.ncrst'
    246 
    247 
    248 def epochSimIndexes(simlist):
    249     """ Finds the simulation indexes for each epoch.

...........................................................................
/home/je714/htmd/htmd/molecule/molecule.py in write(self=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst', sel=None, type=None)
   1144             src.filter(sel, _logger=False)
   1145 
   1146         if type in _WRITERS:
   1147             ext = type
   1148         if ext in _WRITERS:
-> 1149             _WRITERS[ext](src, filename)
        _WRITERS = {'binpos': <function MDTRAJwrite>, 'coor': <function BINCOORwrite>, 'crd': <function MDTRAJwrite>, 'dcd': <function MDTRAJwrite>, 'gro': <function GROwrite>, 'h5': <function MDTRAJwrite>, 'lammpstrj': <function MDTRAJwrite>, 'lh5': <function MDTRAJwrite>, 'mdcrd': <function MDTRAJwrite>, 'mol2': <function MOL2write>, ...}
        ext = 'ncrst'
        src = <htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
   1150 
   1151     def empty(self, numAtoms):
   1152         """ Creates an empty molecule of N atoms.
   1153 

...........................................................................
/home/je714/htmd/htmd/molecule/writers.py in MDTRAJwrite(mol=<htmd.molecule.molecule.Molecule object at 0x7f7..._gen1/structure.prmtop
viewname: structure.prmtop, filename='input/e2s1_e1s1p0f607/structure.ncrst')
    467         # traj.xyz = np.swapaxes(np.swapaxes(self.coords, 1, 2), 0, 1) / 10
    468         # traj.time = self.time
    469         # traj.unitcell_lengths = self.box.T / 10
    470         traj.save(filename)
    471     except Exception as e:
--> 472         raise ValueError('MDtraj reader failed for file {} with error "{}"'.format(filename, e))
        filename = 'input/e2s1_e1s1p0f607/structure.ncrst'
        e = undefined
    473 
    474 
    475 _WRITERS = {'psf': PSFwrite,
    476             'pdb': PDBwrite,

ValueError: MDtraj reader failed for file input/e2s1_e1s1p0f607/structure.ncrst with error "invalid index to scalar variable."
___________________________________________________________________________
```

</details>

and when I re-run the `adapt.run()` command, the following shorter Traceback follows:

```
---------------------------------------------------------------------------
NameError                                 Traceback (most recent call last)
<ipython-input-2-45d070e3237b> in <module>()
     14 adapt.filtersel = 'not water and not resname "Na\+" "Cl\-"'
     15 adapt.coorname = 'structure.ncrst'
---> 16 adapt.run()

~/htmd/htmd/adaptive/adaptive.py in run(self)
     97                 # If currently running simulations are lower than nmin start new ones to reach nmax number of sims
     98                 if self._running <= self.nmin and epoch < self.nepochs:
---> 99                     flag = self._algorithm()
    100                     if flag is False:
    101                         self._unsetLock()

~/htmd/htmd/adaptive/adaptiverun.py in _algorithm(self)
    123 
    124     def _algorithm(self):
--> 125         data = self._getData(self._getSimlist())
    126         if not self._checkNFrames(data): return False
    127         self._createMSM(data)

~/htmd/htmd/adaptive/adaptiverun.py in _getData(self, sims)
    157         if self.ticadim > 0:
    158             # tica = TICA(metr, int(max(2, np.ceil(self.ticalag))))  # gianni: without project it was tooooo slow
--> 159             data = metr.project()
    160             data.dropTraj()  # Drop before TICA to avoid broken trajectories
    161             ticalag = int(

~/htmd/htmd/projections/metric.py in project(self)
    139             for proj in self.projectionlist:
    140                 if isinstance(proj, Projection):
--> 141                     proj._precalculate(uqMol)
    142         else:
    143             logger.warning('Cannot calculate description of dimensions due to different topology files for each trajectory.')

~/htmd/htmd/projections/metricdistance.py in _precalculate(self, mol)
     55 
     56     def _precalculate(self, mol):
---> 57         self.precalcsel1 = self._processSelection(mol, self.sel1, self.groupsel1)
     58         self.precalcsel2 = self._processSelection(mol, self.sel2, self.groupsel2)
     59 

~/htmd/htmd/projections/metricdistance.py in _processSelection(self, mol, sel, groupsel)
    162 
    163         if np.sum(sel) == 0:
--> 164             raise NameError('Selection returned 0 atoms')
    165         return sel
    166 

NameError: Selection returned 0 atoms
```

which makes me think it's using the empty `structure.ncrst` file.

In short, is it even possible to do what I'm trying to attempt?

Thanks for any suggestions

**Comment by stefdoerr** [comment]
Yes totally possible and should work. The way you did it is correct by using a filtersel that gives a common subset of atoms. You might need to delete the `filtered` folder if you changed that filtering selection.

Otherwise it must be an issue with MDtraj as a reader which I have not faced yet. If deleting `filtered` does not solve it could you send me a minimal example? (2 traj from different systems called something like e1sX_X in data/ filtered/ input/ should do it)

**Comment by jeiros** [comment]
Erased the `filtered/` directory and still got the same error. I'm sending you an email with the compressed folder since its ~500 MB.

**Comment by stefdoerr** [comment]
Hm it worked fine for me with the data you gave me because you only gave me one system (still crashes, but at the MSM because of not enough eigenvectors but that's because of too little data). You must give me two trajectories from different systems at least

**Comment by jeiros** [comment]
woops, sorry. Sending the second folder in data/ right away.

**Comment by jeiros** [comment]
yeah, it's already under input/ on the first thing that I sent you

**Comment by stefdoerr** [comment]
Ehm, it works for me. Can you try the exact same dataset you gave me? For
me it worked fine and even wrote the next simulations out

**Comment by jeiros** [comment]
Oh, weird. I'm going to run it then on the thing that I sent you to see if it crashes. 😕

**Comment by jeiros** [comment]
So it has worked out fine. Sorry for opening the issue, I'm confused at what was giving me the error in the first place...

**Comment by j3mdamas** [comment]
Please comment again if you can reproduce it. Thanks for reporting!
